core/kernel/api/class.ModelFactory.php

```diff
@@ -16,112 +17,181 @@
-     * @param string $predicate
-     * @param string $object
-     * @param string $lang
+     * Adds a statement to the ontology if it does not exist yet.
+     * @param string|int $modelId
+     * @param string     $subject
+     * @param string     $predicate
+     * @param string     $object
+     * @param string     $lang
+     * @param string     $author
+     * @return bool Was a row added?
      */
-    private function addStatement($modelId, $subject, $predicate, $object, $lang = null) {
-        $result = core_kernel_classes_DbWrapper::singleton()->query(
+    public function addStatement($modelId, $subject, $predicate, $object, $lang = null, $author = self::DEFAULT_AUTHOR)
+    {
+        // Casting values and types.
+        $object = (string)$object;
+        if (is_null($lang)) {
+            $lang = '';
+        }
+
+        // TODO: refactor this to use a triple store abstraction.
+        $result = $this->dbWrapper->query(
             'SELECT count(*) FROM statements WHERE modelid = ? AND subject = ? AND predicate = ? AND object = ? AND l_language = ?',
-            array($modelId, $subject, $predicate, $object, (is_null($lang)) ? '' : $lang)
+            [$modelId, $subject, $predicate, $object, $lang]
         );
-        
-        if (intval($result->fetchColumn()) === 0) {
-            $dbWrapper = core_kernel_classes_DbWrapper::singleton();
-            $date = $dbWrapper->getPlatForm()->getNowExpression();
-
-            $dbWrapper->insert(
-                'statements',
-                array(
-                    'modelid' =>  $modelId,
-                    'subject' =>$subject,
-                    'predicate'=> $predicate,
-                    'object' => $object,
-                    'l_language' => is_null($lang) ? '' : $lang,
-                    'author' => 'http://www.tao.lu/Ontologies/TAO.rdf#installator',
-                    'epoch' => $date
-                )
-            );
+
+        if (intval($result->fetchColumn()) > 0) {
+            return false;
         }
+
+        return (bool)$this->dbWrapper->insert(
+            'statements',
+            $this->prepareStatement($modelId, $subject, $predicate, $object, $lang, is_null($author) ? '' : $author)
+        );
+    }
+
+    /**
+     * Prepares a statement to be inserted in the ontology.
+     * @param string|int $modelId
+     * @param string     $subject
+     * @param string     $predicate
+     * @param string     $object
+     * @param string     $lang
+     * @param string     $author
+     * @return array
+     */
+    abstract public function prepareStatement($modelId, $subject, $predicate, $object, $lang, $author);
+
+    /**
+     * Creates a query for iterator on selected statements.
+     * @param $modelIds
+     * @return string
+     */
+    public function getIteratorQuery($modelIds)
+    {
+        return 'SELECT * FROM statements '
+            . (is_null($modelIds) ? '' : 'WHERE ' . $this->buildModelSqlCondition($modelIds) . ' ')
```

Comment:
i don't think you need the `is_null`, also i think a `sprintf` will be more readable